lib/ansible/modules/storage/netapp/na_santricity_drive_firmware.py

```diff
@@ -0,0 +1,214 @@
+            - When I(upgrade_drives_online==False) stop all I/O before running task.
+        type: bool
+        default: true
+"""
+EXAMPLES = """
+- name: Ensure correct firmware versions
+  na_santricity_drive_firmware:
+    ssid: "{{ eseries_ssid }}"
+    api_url: "{{ eseries_api_url }}"
+    api_username: "{{ eseries_api_username }}"
+    api_password: "{{ eseries_api_password }}"
+    validate_certs: "{{ eseries_validate_certs }}"
+    firmware: "path/to/drive_firmware"
+    wait_for_completion: true
+    ignore_inaccessible_drives: false
+"""
+RETURN = """
+msg:
+    description: Whether any drive firmware was upgraded and whether it is in progress.
+    type: str
+    returned: always
+    sample:
+        { changed: True, upgrade_in_process: True }
+"""
+import re
+
+from time import sleep
+from ansible.module_utils.netapp import NetAppESeriesModule, create_multipart_formdata
+from ansible.module_utils._text import to_native, to_text, to_bytes
+
+
+class NetAppESeriesDriveFirmware(NetAppESeriesModule):
+    WAIT_TIMEOUT_SEC = 60 * 15
+
+    def __init__(self):
+        ansible_options = dict(
+            firmware=dict(type="list", required=True),
+            wait_for_completion=dict(type="bool", default=False),
+            ignore_inaccessible_drives=dict(type="bool", default=False),
+            upgrade_drives_online=dict(type="bool", default=True))
+
+        super(NetAppESeriesDriveFirmware, self).__init__(ansible_options=ansible_options,
+                                                         web_services_version="02.00.0000.0000",
+                                                         supports_check_mode=True)
+
+        args = self.module.params
+        self.firmware_list = args["firmware"]
+        self.wait_for_completion = args["wait_for_completion"]
+        self.ignore_inaccessible_drives = args["ignore_inaccessible_drives"]
+        self.upgrade_drives_online = args["upgrade_drives_online"]
+
+        self.upgrade_list_cache = None
+
+        self.upgrade_required_cache = None
+        self.upgrade_in_progress = False
+        self.drive_info_cache = None
+
+    def upload_firmware(self):
+        """Ensure firmware has been upload prior to upgrade."""
+        for firmware in self.firmware_list:
+            firmware_name = re.split(r"/", firmware)[-1]
+            files = [("file", firmware_name, firmware)]
+            headers, data = create_multipart_formdata(files)
+            try:
+                rc, response = self.request("/files/drive", method="POST", headers=headers, data=data)
+            except Exception as error:
+                self.module.fail_json(msg="Failed to upload drive firmware [%s]. Array [%s]. Error [%s]." % (firmware_name, self.ssid, to_native(error)))
+
+    def upgrade_list(self):
+        """Determine whether firmware is compatible with the specified drives."""
+        if self.upgrade_list_cache is None:
+            self.upgrade_list_cache = list()
```

Comment:
Why isn't this defined as a dict or a flat set with the drive refs? Is there some future-proofing involved here?

Reply:
This method is producing the body for the storage-systems/%s/firmware/drives/initiate-upgrade post request which requires a list of dictionaries each in the form,
 {"filename": "", "driveRefList": [...]}.